lua/starfall/libs_sh/game.lua

```diff
@@ -134,6 +154,89 @@ if SERVER then
 		util.BlastDamage(instance.entity, instance.player, vunwrap(damageOrigin), math.Clamp(damageRadius, 0, 1500), damage)
 	end
 
+	--- Fires a bullet. Bullet made with this function will not have any tracer, you will have to make them yourself.
+	-- @server
+	-- @param number damage The damage dealt by the bullet. (1-100)
+	-- @param number froce The force of the bullets. (0-100)
+	-- @param number distance Maximum distance the bullet can travel.
+	-- @param number hullSize The hull size of the bullet. (0-10)
+	-- @param number num The amount of bullets to fire. (1-5)
+	-- @param string? ammoType The ammunition name. See enum AMMO_TYPE
+	-- @param Vector Dir The fire direction.
+	-- @param Vector Spread The spread, only x and y are needed.
+	-- @param Vector src The position to fire the bullets from.
+	-- @param Entity? ignoreEntity The entity that the bullet will ignore when it will be shot.
+	-- @param function? callback Function to be called with attacker, traceResult after the bullet was fired but before the damage is applied (the callback is called even if no damage is applied).
+	function game_library.bulletDamage(damage, force, distance, hullSize, num, ammoType, dir, spread, src, ignoreEntity, cb)
+		local ent = instance.entity
+		checkpermission(instance, nil, "bullets.fire")
+		checkluatype(damage, TYPE_NUMBER)
+		checkluatype(force, TYPE_NUMBER)
+		checkluatype(distance, TYPE_NUMBER)
+		checkluatype(hullSize, TYPE_NUMBER)
+		checkluatype(num, TYPE_NUMBER)
+
+		if ammoType then
+			checkluatype(ammoType, TYPE_STRING)
+		end
+
+		local callback
+
+		if cb then
+			checkluatype(cb, TYPE_FUNCTION)
+			callback = function(attacker, tr, dmginfo)
+				instance:runFunction(cb, instance.WrapObject(attacker), SF.StructWrapper(instance, tr, "TraceResult"))
+			end
+		end
+
+		local BulletInfo = {
+			Attacker = instance.player,
+			Callback = callback,
+			Damage = math.Clamp(damage, 1, maxBulletDamage:GetInt()),
+			Force = math.Clamp(force, 0, maxBulletForce:GetInt()),
+			Distance = distance,
+			HullSize = math.min(hullSize, maxBulletHull:GetInt()),
+			Num = math.min(num or 1, maxBulletNum:GetInt()),
+			Tracer = 0,
+			AmmoType = (allowedAmmoType[ammoType] and ammoType) or "SMG1",
+			TracerName = "",
+			Dir = vunwrap(dir),
+			Spread = vunwrap(spread),
+			Src = vunwrap(src),
+			IgnoreEntity = ignoreEntity and eunwrap(ignoreEntity),
+		}
+
+		fireBulletsBurst:use(instance.player, BulletInfo.Num)
+		fireBulletsDPSBurst:use(instance.player, BulletInfo.Damage * BulletInfo.Num)
+
+		ent:FireBullets(BulletInfo)
```

Comment:
Has no infinite loop prevention for https://wiki.facepunch.com/gmod/GM:EntityFireBullets and https://wiki.facepunch.com/gmod/GM:PostEntityFireBullets

Reply:
Starfall has protection for this actually

Reply:
Ah, I see now